File: src/main/java/org/wysko/midis2jam2/instrument/family/percussion/drumset/SnareDrum.java
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.wysko.midis2jam2.instrument.family.percussion.drumset;

import com.jme3.math.Quaternion;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.instrument.family.percussive.Stick;
import org.wysko.midis2jam2.midi.MidiNoteOnEvent;
import org.wysko.midis2jam2.world.Axis;

import java.util.List;
import java.util.stream.Collectors;

import static org.wysko.midis2jam2.Midis2jam2.rad;
import static org.wysko.midis2jam2.instrument.family.percussive.Stick.MAX_ANGLE;
import static org.wysko.midis2jam2.instrument.family.percussive.Stick.STRIKE_SPEED;

/**
 * The Snare drum.
 */
public class SnareDrum extends PercussionInstrument {
	
	/**
	 * The list of hits for regular notes.
	 */
	List<MidiNoteOnEvent> regularHits;
	
	/**
	 * The list of hits for side sticks.
	 */
	List<MidiNoteOnEvent> sideHits;
	
	/**
	 * The snare drum.
	 */
	Spatial drum;
	
	/**
	 * Contains the {@link #regularStick}.
	 */
	Node regularStickNode = new Node();
	
	/**
	 * Contains the side stick.
	 */
	Node sideStickNode = new Node();
	
	/**
	 * The stick used for regular hits.
	 */
	private final Spatial regularStick;
	
	/**
	 * Instantiates a new Snare drum.
	 *
	 * @param context the context
	 * @param hits    the hits
	 */
	public SnareDrum(Midis2jam2 context, List<MidiNoteOnEvent> hits) {
		super(context, hits);
		this.regularHits = hits.stream().filter(e -> e.note == 40 || e.note == 38).collect(Collectors.toList());
		this.sideHits = hits.stream().filter(e -> e.note == 37).collect(Collectors.toList());
		drum = context.loadModel("DrumSet_SnareDrum.obj", "DrumShell_Snare.bmp");
		regularStick = context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp");
		Spatial sideStick = context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp");
		regularStickNode.attachChild(regularStick);
		sideStickNode.attachChild(sideStick);
		recoilNode.attachChild(drum);
		recoilNode.attachChild(regularStickNode);
		recoilNode.attachChild(sideStickNode);
		highLevelNode.attachChild(recoilNode);
		highLevelNode.move(-10.9f, 16, -72.5f);
		highLevelNode.rotate(rad(10), 0, rad(-10));
		regularStickNode.rotate(0, rad(80), 0);
		regularStickNode.move(10, 0, 3);
		sideStick.setLocalTranslation(0, 0, -2);
		sideStick.setLocalRotation(new Quaternion().fromAngles(0, rad(-20), 0));
		sideStickNode.setLocalTranslation(-1, 0.4f, 6);
	}
	
	@Override
	public void tick(double time, float delta) {
		var regularStickStatus = Stick.handleStick(context, regularStick, time, delta, regularHits, STRIKE_SPEED, MAX_ANGLE, Axis.X);
		var sideStickStatus = Stick.handleStick(context, sideStickNode, time, delta, sideHits, STRIKE_SPEED, MAX_ANGLE, Axis.X);
		
		var regVel = 0;
		var sideVel = 0;
		if (regularStickStatus.justStruck()) {
			assert regularStickStatus.getStrike() != null;
			regVel = regularStickStatus.getStrike().velocity;
		}
		if (sideStickStatus.justStruck()) {
			assert sideStickStatus.getStrike() != null;
			sideVel = (int) (sideStickStatus.getStrike().velocity * 0.5);
		}
		int velocity = Math.max(regVel, sideVel);
		PercussionInstrument.recoilDrum(recoilNode, velocity != 0, velocity, delta);
	}
}

File: src/main/java/org/wysko/midis2jam2/instrument/family/percussion/drumset/SnareDrum.kt
/*
 * Copyright (C) 2021 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussion.drumset

import com.jme3.math.Quaternion
import com.jme3.scene.Node
import com.jme3.scene.Spatial
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.family.percussion.Retexturable
import org.wysko.midis2jam2.instrument.family.percussion.RetextureType
import org.wysko.midis2jam2.instrument.family.percussive.Stick
import org.wysko.midis2jam2.midi.Midi
import org.wysko.midis2jam2.midi.MidiNoteOnEvent
import org.wysko.midis2jam2.util.Utils.rad
import org.wysko.midis2jam2.world.Axis

/** The Snare drum. */
class SnareDrum(context: Midis2jam2, hits: MutableList<MidiNoteOnEvent>) : PercussionInstrument(context, hits),
    Retexturable {

    /** The list of hits for regular notes. */
    private val regularHits: MutableList<MidiNoteOnEvent> =
        hits.filter { it.note == Midi.ACOUSTIC_SNARE || it.note == Midi.ELECTRIC_SNARE } as MutableList<MidiNoteOnEvent>

    /** The list of hits for side sticks. */
    private val sideHits: MutableList<MidiNoteOnEvent> =
        hits.filter { it.note == Midi.SIDE_STICK } as MutableList<MidiNoteOnEvent>

    /** Contains the side stick. */
    private val sideStickNode = Node()

    /** The stick used for regular hits. */
    private val regularStick: Spatial
    override fun tick(time: Double, delta: Float) {
        val regularStickStatus = Stick.handleStick(
            context,
            regularStick,
            time,
            delta,
            regularHits,
            Stick.STRIKE_SPEED,
            Stick.MAX_ANGLE,
            Axis.X
        )
        val sideStickStatus = Stick.handleStick(
            context,
            sideStickNode,
            time,
            delta,
            sideHits,
            Stick.STRIKE_SPEED,
            Stick.MAX_ANGLE,
            Axis.X
        )
        var regVel = 0
        var sideVel = 0
        if (regularStickStatus.justStruck()) {
            regVel = (regularStickStatus.strike ?: return).velocity
        }
        if (sideStickStatus.justStruck()) {
            sideVel = ((sideStickStatus.strike ?: return).velocity * 0.5).toInt()
        }
        val velocity = regVel.coerceAtLeast(sideVel)
        recoilDrum(recoilNode, velocity != 0, velocity, delta)
    }

    /** The geometry of the snare drum. */
    val drum: Spatial = context.loadModel("DrumSet_SnareDrum.obj", "DrumShell_Snare.bmp")

    init {
        drum.apply { recoilNode.attachChild(this) }
        regularStick = context.loadModel("DrumSet_Stick.obj", "StickSkin.bmp")

        val sideStick = regularStick.clone().apply {
            sideStickNode.attachChild(this)
        }

        val regularStickNode: Node = Node().apply {
            attachChild(regularStick)
        }

        recoilNode.run {
            attachChild(regularStickNode)
            attachChild(sideStickNode)
        }

        highLevelNode.run {
            attachChild(recoilNode)
            move(-10.9f, 16f, -72.5f)
            rotate(rad(10.0), 0f, rad(-10.0))
        }

        regularStickNode.run {
            rotate(0f, rad(80.0), 0f)
            move(10f, 0f, 3f)
        }

        sideStick.run {
            setLocalTranslation(0f, 0f, -2f)
            localRotation = Quaternion().fromAngles(0f, rad(-20.0), 0f)
        }

        sideStickNode.setLocalTranslation(-1f, 0.4f, 6f)
    }

    override fun drum(): Spatial = drum
    override fun retextureType(): RetextureType = RetextureType.SNARE
}